graph/articulationpoints.go

```diff
@@ -0,0 +1,64 @@
+package graph
+
+import "github.com/TheAlgorithms/Go/math/min"
+
+type apHelper struct {
+	is_ap              []bool
+	visited            []bool
+	child_cnt          []int
+	discovery_time     []int
+	earliest_discovery []int
+}
+
+// ArticulationPoint is a function to identify articulation points in a graph.
+// The function takes the graph as an argument and returns a boolean slice which indicates whether a vertex is an articulation point or not.
+// Worst Case Time Complexity: O(|V| + |E|)
+// Auxiliary Space: O(|V|)
+// reference: https://en.wikipedia.org/wiki/Biconnected_component and https://cptalks.quora.com/Cut-Vertex-Articulation-point
+func ArticulationPoint(graph Graph) []bool {
+	time := 0 // time variable to keep track of the time of discovery_time of a vertex
+
+	//initialize all the variables
+	apHelperInstance := &apHelper{
+		is_ap:              make([]bool, graph.vertices),
+		visited:            make([]bool, graph.vertices),
+		child_cnt:          make([]int, graph.vertices),
+		discovery_time:     make([]int, graph.vertices), // integer slice to store the discovery time of a vertex as we traverse the graph in a depth first manner
+		earliest_discovery: make([]int, graph.vertices), // integer slice to store the earliest discovered vertex reachable from a vertex
+	}
+	articulationPointHelper(apHelperInstance, 0, -1, &time, graph)
+
+	if apHelperInstance.child_cnt[0] == 1 {
+		apHelperInstance.is_ap[0] = false // if the root has only one child, it is not an articulation point
+	}
+
+	return apHelperInstance.is_ap
+}
+
+// articulationPointHelper is a recursive function to traverse the graph and mark articulation points.
+// Based on the depth first search transversal of the graph, however modified to keep track
+// and update the child_cnt, discovery_time and earliest_discovery slices defined above
+func articulationPointHelper(apHelperInstance *apHelper, vertex int, parent int, time *int, graph Graph) {
+	apHelperInstance.visited[vertex] = true
+
+	apHelperInstance.discovery_time[vertex] = *time                                       // Mark the time of discovery of a vertex
+	apHelperInstance.earliest_discovery[vertex] = apHelperInstance.discovery_time[vertex] // set the earliest discovery time to the discovered time
+	*time++                                                                               // increment the time
+
+	for next_vertex := range graph.edges[vertex] {
+		if next_vertex == parent {
+			continue
+		}
+
+		if apHelperInstance.visited[next_vertex] { // if the vertex is already visited, then it is a back edge and manipulate earliest discovery time accordingly
+			apHelperInstance.earliest_discovery[vertex] = min.Int(apHelperInstance.earliest_discovery[vertex], apHelperInstance.discovery_time[next_vertex])
+		} else {
```

Comment:
Do you really need this else? I remember I posted a better way to do this block previously.

Reply:
Here is the [link](https://github.com/TheAlgorithms/Go/pull/452#discussion_r773512443) to that recommendation.

Reply:
You referenced to use `continue` after the `if` block so that the "now" `else` block is not executed. However, I believe the `if-else` block makes more sense here, `if visited do something; else do someotherthing` and therefore kept it there.

Reply:
You may think it makes sense and it is usually the case in other languages but in Go we usually return early. This is a style of writing that is seen everywhere in standard library and is preferred by all experienced devs (IMO). This repo is for Go learners and it should stick with Go's style of writing code and not just copy and/or transliterate other language code.

Reply:
[here](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88) is a link to explain what I'm talking about above. You may disagree with this and that's your right but I would still argue that experienced go devs would frown at `else` statements (I have seen that happen a lot in interviews) and since this is the kind of repo that junior devs would be looking at for thier interview prep, I think we're should stick with idiomatic Go code in this repo.

Read the link for reference. Almost all go devs are aware of this "line of sight" and recommend it to everyone.

This is not a hard and fast take btw, in fact sometimes you can't get rid of the else statements but since in this case you can then I think you should.

Reply:
Understood, thanks for explaining the rationale behind this, I was unaware. Implemented this in the latest commit.